torchbenchmark/score/compute_score.py

```diff
@@ -7,32 +7,137 @@
-    target = config['target']
-    score = 0.0
-    weight_sum = 0.0
-    for name in config['benchmarks']:
-        cfg = config['benchmarks'][name]
-        weight, norm = cfg['weight'], cfg['norm']
-        weight_sum += weight
-        measured_mean = [b['stats']['mean'] for b in data['benchmarks'] if b['name'] == name]
-        assert len(measured_mean) == 1, f"Missing data for {name}, unable to compute score"
-        measured_mean = measured_mean[0]
-        if fake_data is not None and name in fake_data:
-            # used for sanity checks on the sensitivity of the score metric
-            measured_mean = fake_data[name]
-        benchmark_score = weight * math.log(norm / measured_mean)
-        # print(f"{name}: {benchmark_score}")
-        score += benchmark_score
-
-    score = target * math.exp(score)
-    assert abs(weight_sum - 1.0) < 1e-6, f"Bad configuration, weights don't sum to 1, but {weight_sum}"
-    return score
+def _get_model_task(model_name):
+    """
+    Helper function which extracts the task the model belongs to
+    by iterating over the Model attributes.
+    """
+    MODEL_PATH="../../models"
+    p = Path(__file__).parent.joinpath(MODEL_PATH + model_name)
+    if(p):
+        module = importlib.import_module(f'torchbenchmark.models.{model_name}', package=__name__)
+        Model = getattr(module, 'Model')
+    return Model.task.value
 
+class TorchBenchScore:
+    def __init__(self, spec=SPEC_FILE_DEFAULT, target=TARGET_SCORE_DEFAULT, ref_data=None):
+        self.spec = spec
+        self.target = target
+        self.ref_data = ref_data
+        self.weights = None
+        self.norm = None
+
+        # Setup weights and benchmark norms
+        self._setup_weights()
+        self._setup_benchmark_norms()
+
+    def _setup_weights(self):
+        """
+        Calculates the static benchmark weights by iterating the spec
+        file and constructs a dictionary with (key, value) pair
+        is (task, weight_for_benchmark_per_task)
+        """
+        # Load the spec file
+        if Path(self.spec).exists():
+            with open(self.spec) as spec_file:
+                spec = yaml.full_load(spec_file)
+        else:
+            raise ValueError("File {file} doesn't exist".format(file=self.spec))
+
+        self.weights = defaultdict(float)
+        category_spec = spec['hierarchy']['model']
+        domain_weight = 1.0/ len(category_spec)
+        for domain in category_spec:
+            tasks = category_spec[domain]
+            task_weight = 1.0 / len(tasks)
+            for task in tasks:
+                benchmarks = tasks[task]
+                benchmark_weight = 1.0 / len(benchmarks)
+                self.weights[task] = domain_weight * task_weight * benchmark_weight
+
+    def _setup_benchmark_norms(self):
+        """
+        Helper function which gets the normalization values per benchmark
+        by going through the reference data file.
+        """
+        if not self.ref_data: return
```

Comment:
why is it valid to not have ref_data?  shouldn't this be an error and then you don't need the elif?

Reply:
This is to handle the case, when no ref_data is provided in the init. In this case we delay the construction of norm till compute_score is called and ref_data is then initialized  to data.